File: lua/builtin/treesitter.lua
local M = {}

function M.preset()
  return {
    ---Treesitters to be installed.
    ensure_installed = {
      "lua",
    },
    highlight = {
      enable = true,
    },
    incremental_selection = {
      enable = true,
    },
    indent = {
      enable = true,
    },
    context_commentstring = {
      enable = true,
    },
  }
end

---@param minv MiNV
function M.setup(minv)
  local utils = require("utils")

  local function pre_hook(ctx)
    local comment_utils = require("Comment.utils")
    local cmstring_utils = require("ts_context_commentstring.utils")

    local key = "__default"
    if ctx.ctype ~= comment_utils.ctype.line then
      key = "__multiline"
    end

    local location = nil
    if ctx.ctype == comment_utils.ctype.block then
      location = cmstring_utils.get_cursor_location()
    elseif ctx.cmotion == comment_utils.cmotion.v or ctx.cmotion == comment_utils.cmotion.V then
      location = cmstring_utils.get_visual_start_location()
    end

    return require("ts_context_commentstring.internal").calculate_commentstring({
      key = key,
      location = location,
    })
  end

  local keys = {
    toggle = {
      line = { utils.register_key(), "Toggle line comments" },
      block = { utils.register_key(), "Toggle block comments" },
    },
    oplead = {
      line = { utils.register_key(), "Toggle line comments" },
      block = { utils.register_key(), "Toggle block comments" },
    },
    insert = {
      above = { utils.register_key(), "Insert comment above" },
      below = { utils.register_key(), "Insert comment below" },
      eol = { utils.register_key(), "Insert comment at EOL" },
    },
  }

  ---@param group string
  local function make_mapping(group)
    local mapping = {}
    for k, v in pairs(keys[group]) do
      mapping[k] = v[1]
    end
    return mapping
  end

  ---@vararg string[]
  local function make_source(...)
    local source = {}
    for _, group in ipairs({ ... }) do
      for k, v in pairs(keys[group]) do
        source[string.format("comment.%s_%s", group, k)] = { v[1], v[2], noremap = false }
      end
    end
    return source
  end

  -- Setup treesitter.
  require("nvim-treesitter.configs").setup(minv.builtin.treesitter)

  -- Setup comments.
  require("Comment").setup(vim.tbl_extend("force", minv.builtin.comment, {
    pre_hook = pre_hook,
    toggler = make_mapping("toggle"),
    opleader = make_mapping("oplead"),
    extra = make_mapping("insert"),
    mappings = {
      basic = true,
      extra = true,
      extended = false,
    },
  }))

  -- Set keybindings.
  minv.keybindings.n:apply(false, make_source("toggle", "oplead", "insert"))
  minv.keybindings.x:apply(false, make_source("oplead"), { mode = "x" })
end

return M

File: src/minv/plugins/treesitter.ts
import {
  apply_mappings,
  Source,
  mkPlugKeySrc,
  apply_extra,
} from "../keybindings";
import { PRESETS } from "../presets";
import { deep_merge } from "../utils";

const MAPPINGS = {
  toggle: mkPlugKeySrc({
    "comment.toggle_line": "Toggle line comments",
    "comment.toggle_block": "Toggle block comments",
  }),
  oplead: mkPlugKeySrc({
    "comment.oplead_line": "Toggle line comments",
    "comment.oplead_block": "Toggle block comments",
  }),
  insert: mkPlugKeySrc({
    "comment.insert_above": "Insert comments above",
    "comment.insert_below": "Insert comments below",
    "comment.insert_eol": "Insert comments eol",
  }),
};

function mkMappingsForComment(
  this: void,
  target: keyof typeof MAPPINGS
): { [k: string]: string } {
  const trim = "comment._".length + target.length + 1;
  const mappings: { [k: string]: string } = {};
  for (const [name, src] of pairs(
    MAPPINGS[target] as { [k: string]: Source }
  )) {
    mappings[string.sub(name as string, trim)] = src.cmd as string;
  }
  return mappings as any;
}

const COMMENT_DEFAULT = {
  toggler: mkMappingsForComment("toggle"),
  opleader: mkMappingsForComment("oplead"),
  extra: mkMappingsForComment("insert"),
  mappings: {
    basic: true,
    extra: true,
    extended: false,
  },
};

export function setup_treesitter(this: void) {
  (require("nvim-treesitter.configs") as AnyTbl).setup(PRESETS.treesitter);
}

export function setup_comment(this: void) {
  const pre_hook = (
    require("ts_context_commentstring.integrations.comment_nvim") as AnyTbl
  ).create_pre_hook();
  (require("Comment") as AnyTbl).setup(
    deep_merge("keep", { pre_hook }, COMMENT_DEFAULT, PRESETS.comment)
  );
  apply_mappings(MAPPINGS.toggle, { noremap: false });
  apply_mappings(MAPPINGS.oplead, { mode: "x", noremap: false });
  apply_mappings(MAPPINGS.oplead, { noremap: false });
  apply_mappings(MAPPINGS.insert, { noremap: false });
  apply_extra("comment.extra", { mode: "n" });
}

export function setup_surround(this: void) {
  (require("nvim-surround") as AnyTbl).setup(PRESETS.surround);
}
